Add PeerDAS RPC import boilerplate

dapplion opened the pull request:
## Issue Addressed

Continue porting network changes of `das` branch. Extends https://github.com/sigp/lighthouse/pull/6224

Part of 
- https://github.com/sigp/lighthouse/issues/4983

## Proposed Changes

- Add a new `Work:: RpcCustodyColumn` event
- Wire process function with importing columns into da_checker
- Introduce new type `DataColumnsSameBlock` as downstream code assumes that all submitted columns are in the same block. The implementation in `das` just assumes this to be the case, but this PR makes this assumption explicit via a new type. I changed the gossip import method from accepting N columns to exactly 1, so we can assert that the single column is always part of the same block.

dapplion committed:
Add PeerDAS RPC import boilerplate

dapplion commented on beacon_node/network/src/network_beacon_processor/mod.rs:
```diff
@@ -476,6 +477,24 @@ impl<T: BeaconChainTypes> NetworkBeaconProcessor<T> {
 
+    /// Create a new `Work` event for some custody columns. `process_rpc_custody_columns` reports
+    /// the result back to sync.
+    pub fn send_rpc_custody_columns(
```
Future PR touching sync will consume this function

jimmygchen commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -192,20 +212,18 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
-        slot: Slot,
-        block_root: Hash256,
-        gossip_data_columns: Vec<GossipVerifiedDataColumn<T>>,
+        gossip_data_column: GossipVerifiedDataColumn<T>,
```
I changed this in #5806 to avoid reconstruction for proposer nodes. Is this scenario handled?

dapplion replied:
Right, thanks for pointing that out I wasn't aware. `publish_block` also ensures that all columns are from the same block. Do you think we need the `*SameBlock` type here too? Switching to a single item was just a way to get around having to create a `GossipVerifiedDataColumnsSameBlock` or similar.

jimmygchen replied:
I think it was clear that these are from the same block given the `block_root` and `slot` parameters. Is it necessary to introduce a new type? as we always either pass a single data column around, or pass a list of data columns for the same block. 

Could we maybe delay these kind of refactor in separate future PRs. Feels like it makes conga line PRs difficult to manage (we'll also have to re-work all subsequent PRs)

realbigsean commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -2968,23 +2970,13 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
 
     /// Cache the data columns in the processing cache, process it, then evict it from the cache if it was
```
```suggestion
    /// Cache the data column in the processing cache, process it, then evict it from the cache if it was
```

realbigsean commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -197,20 +217,18 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
+
+        // Will never error as there's exactly one column
+        let custody_columns_same_block = KzgVerifiedDataColumnsSameBlock::new(vec![custody_column])
+            .expect("exactly one column is always in same block");
```
I think it'd be better to make a new infallible constructor in `KzgVerifiedDataColumnsSameBlock` that takes a single column as a param

realbigsean commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -221,34 +222,91 @@ impl<E: EthSpec> KzgVerifiedDataColumn<E> {
+        let signed_block_header = first_column.signed_block_header.clone();
+        for column in columns.iter().skip(1) {
+            if column.signed_block_header != signed_block_header {
+                return Err("no same block");
```
`not same block` or maybe more detailed`mismatched data column headers` or something

realbigsean commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -221,34 +222,91 @@ impl<E: EthSpec> KzgVerifiedDataColumn<E> {
+        let signed_block_header = first_column.as_data_column().signed_block_header.clone();
+        for column in columns.iter().skip(1) {
+            if column.as_data_column().signed_block_header != signed_block_header {
+                return Err("no same block");
```
same as above

realbigsean reviewed (commented):
Just some small feedback but looks good!

dapplion committed:
revert refactor

dapplion commented:
@jimmygchen I have reverted the refactor of the *SameBlock types. This PR just extends same style as blobs

Also rebased on unstable to clean the commits :)

realbigsean commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -181,6 +183,38 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
 
+    /// Put a list of custody columns received via RPC into the availability cache. This performs KZG
+    /// verification on the blobs in the list.
+    #[allow(clippy::type_complexity)]
```
```suggestion
```

i think this is not necessary

realbigsean reviewed (approved):
Looks like my feedback was related to the refactor, just had one nit!

dapplion committed:
Remove allow

realbigsean commented:
@mergify queue
